compiler/damlc/daml-compiler/src/DA/Daml/Compiler/DataDependencies.hs

```diff
@@ -1305,6 +1305,7 @@ isDFunBody = \case
     LF.ETyLam _ body -> isDFunBody body
     LF.ETmLam _ body -> isDFunBody body
     LF.EStructCon _ -> True
+    LF.EUnit -> True
```

Comment:
What happens if I define stuff of type `()` in Daml? Will that now reconstruct them as typeclass instances breaking other things?

Reply:
that's a good point, thank you. 

I checked, and it seems we're not at risk of that since:

* `isDFunBody` is only used by `instanceDecls` (through `shouldExposeInstance`) _after_ the binding's signature is matched as `Just dfunSig <- [getDFunSig dval]`
* In turn, `getDFunSig` returns a `Just` only for `LF.TSynApp` (possibly nested under `->`s and `forall`s)
* `()` is not an `LF.TSynApp`, so a definition with `: ()` will not be picked up as an instance.
* Now you must be thinking you could go around that by defining a type synonym that resolves to `()`, but there's no way to do that since [`LFConversion` skips `()` type synonyms](https://github.com/digital-asset/daml/blob/main/compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs#L731-L733).
* But what about a synonym of `()` nested under `->`s or `forall`s? That's not a problem either, since Daml type synonyms [are fully expanded in `LFConversion`](https://github.com/digital-asset/daml/blob/main/compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs#L2196-L2197), while LF type synonyms are [only generated for `ClassFlavour`ed type constructors](https://github.com/digital-asset/daml/blob/main/compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs#L2202-L2204) (making the point above moot)

Reply:
thanks for checking! Maybe worth including those cases in the test?

Reply:
Thanks for making me check, it really could have broken other things.

We do have a test checking that we skip `()` type synonyms, see `MyUnit` and `MyConstUnit` [here](https://github.com/digital-asset/daml/blob/main/compiler/damlc/tests/src/DA/Test/DataDependencies.hs#L446-L499).

For the second case (synonyms of `()` nested under `->`s or `forall`s), I thought about using a similar approach to the test above, having a fake instance (`()`-valued function) in the data-dependency and checking that the main project fails with a missing instance error. However, for that to work either
1. the type of the fake instance would have to be a synonym of an actual class, but the GHC typechecker doesn't allow bindings with a `Constraint`-kinded type anyway.
2. to work around 1, we could make the class itself fake (and thus `*`-kinded), but `convertClassDef` only cares about types where `tyConClass_maybe` returns a `Just`.